File: scripts/lib/docs-validate.py
#!/usr/bin/env python3
"""Validate linked spec/plan docs for workflow handoff/implement gates."""
from __future__ import annotations

import json
import os
import re
import subprocess
import sys
from pathlib import Path

BRANCH_RE = re.compile(
    r"^\s*\*+Branch:\*+\s*`?((?:feature|bugfix)/[^`\s|]+)`?\s*$",
    re.I | re.M,
)
SPEC_LINK_RE = re.compile(r"^\s*\*+Spec:\*+\s*(?:`([^`]+)`|(\S+))\s*$", re.I | re.M)
TASK_RE = re.compile(r"^###\s+Task\s+(\d+):\s*(.*)$", re.I)


def err(code: str, message: str, path: str | None = None) -> dict:
    item: dict[str, str] = {"code": code, "message": message}
    if path:
        item["path"] = path
    return item


def read_branch(text: str, label: str) -> tuple[str | None, dict | None]:
    match = BRANCH_RE.search(text)
    if not match:
        return None, err("missing_branch", f"**Branch:** feature/* or bugfix/* required in {label}")
    return match.group(1).strip().strip("`"), None


def scan_task_headings(plan_text: str) -> tuple[list[int], list[str], dict | None]:
    ids: list[int] = []
    titles: list[str] = []
    in_fence = False
    for line in plan_text.splitlines():
        if line.startswith("```"):
            in_fence = not in_fence
            continue
        if in_fence:
            continue
        match = TASK_RE.match(line)
        if match:
            ids.append(int(match.group(1)))
            titles.append(match.group(2).strip())
    if not ids:
        return ids, titles, err("task_order", "no ### Task N sections found outside fences")
    expected = list(range(1, len(ids) + 1))
    if sorted(ids) != expected or len(set(ids)) != len(ids):
        return ids, titles, err(
            "task_order",
            f"task headings must be contiguous from 1..{len(ids)}; found {ids}",
        )
    return ids, titles, None


def main() -> int:
    if len(sys.argv) < 4:
        print(json.dumps({"ok": False, "errors": [err("usage", "spec_path plan_path parse_script required")]}))
        return 1

    spec_arg, plan_arg, parser = sys.argv[1], sys.argv[2], sys.argv[3]
    cwd = Path.cwd()
    spec_path = Path(spec_arg) if os.path.isabs(spec_arg) else cwd / spec_arg
    plan_path = Path(plan_arg) if os.path.isabs(plan_arg) else cwd / plan_arg
    errors: list[dict] = []

    if not spec_path.is_file():
        errors.append(err("missing_file", f"spec not found: {spec_arg}", spec_arg))
    if not plan_path.is_file():
        errors.append(err("missing_file", f"plan not found: {plan_arg}", plan_arg))
    if errors:
        print(json.dumps({"ok": False, "errors": errors}))
        return 1

    spec_text = spec_path.read_text(encoding="utf-8")
    plan_text = plan_path.read_text(encoding="utf-8")

    spec_branch, spec_err = read_branch(spec_text, "spec")
    if spec_err:
        errors.append(spec_err)
    plan_branch, plan_err = read_branch(plan_text, "plan")
    if plan_err:
        errors.append(plan_err)

    link_match = SPEC_LINK_RE.search(plan_text)
    if not link_match:
        errors.append(err("missing_spec_link", "**Spec:** link required in plan", str(plan_arg)))
    else:
        linked = (link_match.group(1) or link_match.group(2) or "").strip()
        linked_path = Path(linked) if os.path.isabs(linked) else cwd / linked
        try:
            if linked_path.resolve() != spec_path.resolve():
                errors.append(
                    err(
                        "spec_mismatch",
                        f"plan **Spec:** {linked} does not match spec_path {spec_arg}",
                        str(plan_arg),
                    )
                )
        except OSError:
            errors.append(err("spec_mismatch", f"cannot resolve plan **Spec:** {linked}", str(plan_arg)))

    if spec_branch and plan_branch and spec_branch != plan_branch:
        errors.append(
            err(
                "branch_mismatch",
                f"spec branch {spec_branch!r} != plan branch {plan_branch!r}",
                str(plan_arg),
            )
        )

    _, _, task_err = scan_task_headings(plan_text)
    if task_err:
        errors.append(task_err)

    if errors:
        print(json.dumps({"ok": False, "errors": errors}))
        return 1

    parse = subprocess.run(
        ["bash", parser, str(plan_path), "--format=json"],
        cwd=str(cwd),
        capture_output=True,
        text=True,
        check=False,
    )
    if parse.returncode != 0:
        message = (parse.stderr or parse.stdout or "parse-plan-tasks failed").strip()
        print(json.dumps({"ok": False, "errors": [err("parse_failed", message, str(plan_arg))]}))
        return 1

    try:
        parsed = json.loads(parse.stdout.strip())
    except json.JSONDecodeError:
        print(json.dumps({"ok": False, "errors": [err("parse_failed", "invalid JSON from parse-plan-tasks", str(plan_arg))]}))
        return 1

    parsed_tasks = parsed.get("tasks") or []
    heading_ids, heading_titles, task_err = scan_task_headings(plan_text)
    if task_err:
        print(json.dumps({"ok": False, "errors": [task_err]}))
        return 1

    if len(parsed_tasks) != len(heading_ids):
        print(
            json.dumps(
                {
                    "ok": False,
                    "errors": [
                        err(
                            "task_order",
                            f"parse-plan-tasks count {len(parsed_tasks)} != heading count {len(heading_ids)}",
                            str(plan_arg),
                        )
                    ],
                }
            )
        )
        return 1

    for i, task in enumerate(parsed_tasks):
        if str(task.get("id")) != str(heading_ids[i]):
            print(
                json.dumps(
                    {
                        "ok": False,
                        "errors": [
                            err(
                                "task_order",
                                f"task id mismatch at position {i + 1}",
                                str(plan_arg),
                            )
                        ],
                    }
                )
            )
            return 1
        if (task.get("title") or "").strip() != heading_titles[i]:
            print(
                json.dumps(
                    {
                        "ok": False,
                        "errors": [
                            err(
                                "task_order",
                                f"task title mismatch for Task {heading_ids[i]}",
                                str(plan_arg),
                            )
                        ],
                    }
                )
            )
            return 1

    rel_spec = spec_arg if not os.path.isabs(spec_arg) else os.path.relpath(spec_path, cwd)
    rel_plan = plan_arg if not os.path.isabs(plan_arg) else os.path.relpath(plan_path, cwd)
    print(
        json.dumps(
            {
                "ok": True,
                "spec": rel_spec,
                "plan": rel_plan,
                "branch": spec_branch,
                "task_count": len(parsed_tasks),
            }
        )
    )
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: src/core/docs-validate.ts
import { readFileSync } from "node:fs";
import path from "node:path";

export type DocError = { code: string; message: string; path?: string };

const BRANCH_RE = /^\s*\*+Branch:\*+\s*`?((?:feature|bugfix)\/[^`\s|]+)`?\s*$/im;
const SPEC_LINK_RE = /^\s*\*+Spec:\*+\s*(?:`([^`]+)`|(\S+))\s*$/im;
const TASK_RE = /^###\s+Task\s+(\d+):\s*(.*)$/i;

const err = (code: string, message: string, path?: string): DocError => {
  const item: DocError = { code, message };
  if (path) item.path = path;
  return item;
};

const failValidate = (
  errors: DocError[],
): { ok: false; errors: DocError[]; error: string } => ({
  ok: false,
  errors,
  error: errors.map((e) => e.message).filter(Boolean).join("; ") || "docs validation failed",
});

const readBranch = (text: string, label: string): [string | null, DocError | null] => {
  const match = text.match(BRANCH_RE);
  if (!match) return [null, err("missing_branch", `**Branch:** feature/* or bugfix/* required in ${label}`)];
  return [match[1].trim().replace(/`/g, ""), null];
};

const scanTaskHeadings = (planText: string): [number[], string[], DocError | null] => {
  const ids: number[] = [];
  const titles: string[] = [];
  let inFence = false;
  for (const line of planText.split("\n")) {
    if (line.startsWith("```")) { inFence = !inFence; continue; }
    if (inFence) continue;
    const match = line.match(TASK_RE);
    if (match) { ids.push(Number(match[1])); titles.push(match[2].trim()); }
  }
  if (ids.length === 0) return [ids, titles, err("task_order", "no ### Task N sections found outside fences")];
  const expected = ids.map((_, i) => i + 1);
  const sorted = [...ids].sort((a, b) => a - b);
  if (JSON.stringify(sorted) !== JSON.stringify(expected) || new Set(ids).size !== ids.length) {
    return [ids, titles, err("task_order", `task headings must be contiguous from 1..${ids.length}; found ${ids}`)];
  }
  return [ids, titles, null];
};

// Port of scripts/lib/parse-plan-tasks.sh (JSON mode)
export const parseTasksFromPlan = (planText: string): { id: number; title: string; section_text: string }[] => {
  const tasks: { id: number; title: string; section_text: string }[] = [];
  let current: { id: number; title: string; body: string[] } | null = null;
  let inFence = false;
  for (const line of planText.split("\n")) {
    if (line.startsWith("```")) { inFence = !inFence; continue; }
    if (inFence) continue;
    const match = line.match(TASK_RE);
    if (match) {
      if (current) tasks.push({ id: current.id, title: current.title, section_text: current.body.join("\n") });
      current = { id: Number(match[1]), title: match[2].trim(), body: [] };
    } else if (current) {
      current.body.push(line);
    }
  }
  if (current) tasks.push({ id: current.id, title: current.title, section_text: current.body.join("\n") });
  return tasks;
};

export const docsValidate = ({
  spec_path,
  plan_path,
  workspace_root,
}: {
  spec_path: string;
  plan_path: string;
  workspace_root: string;
}):
  | { ok: true; spec: string; plan: string; branch: string; task_count: number; quality: QualityFinding[] }
  | { ok: false; errors: DocError[]; error: string } => {
  const cwd = path.resolve(workspace_root);
  const specAbs = path.isAbsolute(spec_path) ? spec_path : path.join(cwd, spec_path);
  const planAbs = path.isAbsolute(plan_path) ? plan_path : path.join(cwd, plan_path);
  const errors: DocError[] = [];

  const read = (p: string): string | null => {
    try { return readFileSync(p, "utf8"); } catch { return null; }
  };

  const specText = read(specAbs);
  const planText = read(planAbs);
  if (specText === null) errors.push(err("missing_file", `spec not found: ${spec_path}`, spec_path));
  if (planText === null) errors.push(err("missing_file", `plan not found: ${plan_path}`, plan_path));
  if (errors.length) return failValidate(errors);

  const [specBranch, specErr] = readBranch(specText!, "spec");
  if (specErr) errors.push(specErr);
  const [planBranch, planErr] = readBranch(planText!, "plan");
  if (planErr) errors.push(planErr);

  const linkMatch = planText!.match(SPEC_LINK_RE);
  if (!linkMatch) {
    errors.push(err("missing_spec_link", "**Spec:** link required in plan", plan_path));
  } else {
    const linked = (linkMatch[1] ?? linkMatch[2] ?? "").trim();
    const linkedAbs = path.isAbsolute(linked) ? linked : path.join(cwd, linked);
    if (path.resolve(linkedAbs) !== path.resolve(specAbs)) {
      errors.push(err("spec_mismatch", `plan **Spec:** ${linked} does not match spec_path ${spec_path}`, plan_path));
    }
  }

  if (specBranch && planBranch && specBranch !== planBranch) {
    errors.push(err("branch_mismatch", `spec branch ${JSON.stringify(specBranch)} != plan branch ${JSON.stringify(planBranch)}`, plan_path));
  }

  const [, , taskErr] = scanTaskHeadings(planText!);
  if (taskErr) errors.push(taskErr);

  if (errors.length) return failValidate(errors);

  const tasks = parseTasksFromPlan(planText!);
  const [headingIds, headingTitles, headingErr] = scanTaskHeadings(planText!);
  if (headingErr) return failValidate([headingErr]);
  if (tasks.length !== headingIds.length) {
    return failValidate([err("task_order", `parse count ${tasks.length} != heading count ${headingIds.length}`, plan_path)]);
  }
  for (let i = 0; i < tasks.length; i++) {
    if (String(tasks[i].id) !== String(headingIds[i]) || tasks[i].title.trim() !== headingTitles[i]) {
      return failValidate([err("task_order", `task mismatch at position ${i + 1}`, plan_path)]);
    }
  }

  const relSpec = path.isAbsolute(spec_path) ? path.relative(cwd, specAbs) : spec_path;
  const relPlan = path.isAbsolute(plan_path) ? path.relative(cwd, planAbs) : plan_path;
  return {
    ok: true,
    spec: relSpec,
    plan: relPlan,
    branch: specBranch!,
    task_count: tasks.length,
    quality: qualitySpec(specText!),
  };
};

export type QualityFinding = {
  code: string;
  message: string;
  severity: "warning" | "hard";
};

const REQUIRED_SECTIONS = [
  "## Context",
  "## Goals",
  "## Non-goals",
  "## Architecture",
  "## Acceptance criteria",
];

const UI_KEYWORDS = [/\bui\b/, /\binterface\b/, /\bscreen\b/, /\bmodal\b/, /\bform\b/, /\bcomponent\b/];
const FLOW_KEYWORDS = [/\bflow\b/, /\bpipeline\b/, /\bsequence\b/, /\bdiagram\b/];
const GLOSSARY_KEYWORDS = [/\bglossary\b/, /\bcontracts?\b/, /\bscope\b/];

const finding = (code: string, message: string, severity: "warning" | "hard"): QualityFinding =>
  ({ code, message, severity });

// Replace fenced code blocks with a single marker line so their content cannot
// satisfy the checks, but the fence itself (and its language) stays detectable.
const stripFences = (text: string): string => {
  const lines = text.split("\n");
  const out: string[] = [];
  let inFence = false;
  for (const line of lines) {
    if (line.startsWith("```")) {
      if (!inFence) out.push(line); // opening fence: keep the ```lang marker
      inFence = !inFence;
      continue;
    }
    if (!inFence) out.push(line);
  }
  return out.join("\n");
};

export const qualitySpec = (text: string): QualityFinding[] => {
  const findings: QualityFinding[] = [];
  const body = stripFences(text);
  const lower = body.toLowerCase();

  for (const section of REQUIRED_SECTIONS) {
    if (!body.includes(section)) {
      findings.push(finding("missing_section", `required section ${section} missing`, "hard"));
    }
  }

  const hasCa = /^\s*(?:- CA-\d+|CA-\d+[.:])/m.test(body); // M1: ^ with /m covers line starts
  if (!hasCa) {
    findings.push(finding("missing_acceptance_criteria", "no enumerable CA-XX acceptance criteria found", "hard"));
  }

  const hasAsciiFence = /```(?:text|ascii)/.test(body);
  const mentionsUi = UI_KEYWORDS.some((k) => k.test(lower));
  if (mentionsUi && !hasAsciiFence) {
    findings.push(finding("missing_ascii_for_ui", "spec mentions UI but has no ASCII wireframe fence", "warning"));
  }

  const hasMermaid = /```mermaid/.test(body);
  const explicitlyNoFlow = /\bno (?:flow|pipeline|sequence|diagram)\b/.test(lower);
  const mentionsFlow = FLOW_KEYWORDS.some((k) => k.test(lower));
  if (mentionsFlow && !hasMermaid && !explicitlyNoFlow) {
    findings.push(finding("missing_mermaid_for_flow", "spec describes a flow/pipeline/sequence but has no mermaid fence", "warning"));
  }

  const hasTable = /^\s*\|.+\|.+\|/m.test(body);
  const mentionsGlossary = GLOSSARY_KEYWORDS.some((k) => k.test(lower));
  const onlyOutOfScope = /\bout of scope\b/.test(lower) && !/\bglossary\b/.test(lower) && !/\bcontracts?\b/.test(lower);
  if (mentionsGlossary && !hasTable && !onlyOutOfScope) {
    findings.push(finding("missing_table", "spec has glossary/contract/scope content but no markdown table", "warning"));
  }

  return findings;
};
